<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head><title>R: Exploratory Analysis of Genetic and Genomic Data</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>
<h1> Exploratory Analysis of Genetic and Genomic Data
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;adegenet&rsquo; version 2.1.3</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
</ul>

<h2>Help Pages</h2>


<p style="text-align: center;">
<a href="#A">A</a>
<a href="#B">B</a>
<a href="#C">C</a>
<a href="#D">D</a>
<a href="#E">E</a>
<a href="#F">F</a>
<a href="#G">G</a>
<a href="#H">H</a>
<a href="#I">I</a>
<a href="#K">K</a>
<a href="#L">L</a>
<a href="#M">M</a>
<a href="#N">N</a>
<a href="#O">O</a>
<a href="#P">P</a>
<a href="#R">R</a>
<a href="#S">S</a>
<a href="#T">T</a>
<a href="#U">U</a>
<a href="#V">V</a>
<a href="#W">W</a>
<a href="#X">X</a>
<a href="#misc">misc</a>
</p>


<h2><a name="A">-- A --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="ascore.html">a.score</a></td>
<td>Compute and optimize a-score for Discriminant Analysis of Principal Components (DAPC)</td></tr>
<tr><td style="width: 25%;"><a href="strata-methods.html">addStrata</a></td>
<td>Access and manipulate the population strata for genind or genlight objects.</td></tr>
<tr><td style="width: 25%;"><a href="strata-methods.html">addStrata-method</a></td>
<td>Access and manipulate the population strata for genind or genlight objects.</td></tr>
<tr><td style="width: 25%;"><a href="strata-methods.html">addStrata&lt;-</a></td>
<td>Access and manipulate the population strata for genind or genlight objects.</td></tr>
<tr><td style="width: 25%;"><a href="strata-methods.html">addStrata&lt;--method</a></td>
<td>Access and manipulate the population strata for genind or genlight objects.</td></tr>
<tr><td style="width: 25%;"><a href="adegenet.package.html">adegenet</a></td>
<td>The adegenet package</td></tr>
<tr><td style="width: 25%;"><a href="adegenet.package.html">adegenet.package</a></td>
<td>The adegenet package</td></tr>
<tr><td style="width: 25%;"><a href="web.html">adegenetIssues</a></td>
<td>Functions to access online resources for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="servers.html">adegenetServer</a></td>
<td>Web servers for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="web.html">adegenetTutorial</a></td>
<td>Functions to access online resources for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="web.html">adegenetWeb</a></td>
<td>Functions to access online resources for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="AIC.snapclust.html">AIC.snapclust</a></td>
<td>Compute Akaike Information Criterion (AIC) for snapclust</td></tr>
<tr><td style="width: 25%;"><a href="AICc.html">AICc</a></td>
<td>Compute Akaike Information Criterion for small samples (AICc) for snapclust</td></tr>
<tr><td style="width: 25%;"><a href="AICc.html">AICc.snapclust</a></td>
<td>Compute Akaike Information Criterion for small samples (AICc) for snapclust</td></tr>
<tr><td style="width: 25%;"><a href="sequences.html">alignment2genind</a></td>
<td>Importing data from an alignement of sequences to a genind object</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">alleles</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">alleles-method</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">alleles-method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">alleles&lt;-</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">alleles&lt;--method</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">alleles&lt;--method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="auxil.html">any2col</a></td>
<td>Auxiliary functions for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="SNPbin.html">as-method</a></td>
<td>Formal class "SNPbin"</td></tr>
<tr><td style="width: 25%;"><a href="as-methods.html">as-method</a></td>
<td>Converting genind/genpop objects to other classes</td></tr>
<tr><td style="width: 25%;"><a href="as.SNPbin.html">as-method</a></td>
<td>Conversion to class "SNPbin"</td></tr>
<tr><td style="width: 25%;"><a href="as.genlight.html">as-method</a></td>
<td>Conversion to class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">as-method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="as-methods.html">as.data.frame.genind</a></td>
<td>Converting genind/genpop objects to other classes</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">as.data.frame.genlight</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="as-methods.html">as.data.frame.genpop</a></td>
<td>Converting genind/genpop objects to other classes</td></tr>
<tr><td style="width: 25%;"><a href="new.genind.html">as.genind</a></td>
<td>genind constructor</td></tr>
<tr><td style="width: 25%;"><a href="as.genlight.html">as.genlight</a></td>
<td>Conversion to class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="as.genlight.html">as.genlight-method</a></td>
<td>Conversion to class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="new.genpop.html">as.genpop</a></td>
<td>genpop constructor</td></tr>
<tr><td style="width: 25%;"><a href="as-methods.html">as.genpop.genind</a></td>
<td>Converting genind/genpop objects to other classes</td></tr>
<tr><td style="width: 25%;"><a href="haploGen.html">as.igraph.haploGen</a></td>
<td>Simulation of genealogies of haplotypes</td></tr>
<tr><td style="width: 25%;"><a href="seqTrack.html">as.igraph.seqTrack</a></td>
<td>SeqTrack algorithm for reconstructing genealogies</td></tr>
<tr><td style="width: 25%;"><a href="SNPbin.html">as.integer.SNPbin</a></td>
<td>Formal class "SNPbin"</td></tr>
<tr><td style="width: 25%;"><a href="as-methods.html">as.ktab.genind</a></td>
<td>Converting genind/genpop objects to other classes</td></tr>
<tr><td style="width: 25%;"><a href="as-methods.html">as.ktab.genpop</a></td>
<td>Converting genind/genpop objects to other classes</td></tr>
<tr><td style="width: 25%;"><a href="dapc.html">as.lda</a></td>
<td>Discriminant Analysis of Principal Components (DAPC)</td></tr>
<tr><td style="width: 25%;"><a href="dapc.html">as.lda.dapc</a></td>
<td>Discriminant Analysis of Principal Components (DAPC)</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">as.list.genlight</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="as-methods.html">as.matrix.genind</a></td>
<td>Converting genind/genpop objects to other classes</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">as.matrix.genlight</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="as-methods.html">as.matrix.genpop</a></td>
<td>Converting genind/genpop objects to other classes</td></tr>
<tr><td style="width: 25%;"><a href="haploGen.html">as.POSIXct.haploGen</a></td>
<td>Simulation of genealogies of haplotypes</td></tr>
<tr><td style="width: 25%;"><a href="haploGen.html">as.seqTrack.haploGen</a></td>
<td>Simulation of genealogies of haplotypes</td></tr>
<tr><td style="width: 25%;"><a href="as.SNPbin.html">as.SNPbin</a></td>
<td>Conversion to class "SNPbin"</td></tr>
<tr><td style="width: 25%;"><a href="as.SNPbin.html">as.SNPbin-method</a></td>
<td>Conversion to class "SNPbin"</td></tr>
<tr><td style="width: 25%;"><a href="dapcGraphics.html">assignplot</a></td>
<td>Graphics for Discriminant Analysis of Principal Components (DAPC)</td></tr>
<tr><td style="width: 25%;"><a href="auxil.html">azur</a></td>
<td>Auxiliary functions for adegenet</td></tr>
</table>

<h2><a name="B">-- B --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="BIC.snapclust.html">BIC.snapclust</a></td>
<td>Compute Bayesian Information Criterion (BIC) for snapclust</td></tr>
<tr><td style="width: 25%;"><a href="doc_C_routines.html">binIntToBytes</a></td>
<td>Internal C routines</td></tr>
<tr><td style="width: 25%;"><a href="auxil.html">bluepal</a></td>
<td>Auxiliary functions for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="doc_C_routines.html">bytesToBinInt</a></td>
<td>Internal C routines</td></tr>
<tr><td style="width: 25%;"><a href="doc_C_routines.html">bytesToInt</a></td>
<td>Internal C routines</td></tr>
</table>

<h2><a name="C">-- C --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="SNPbin.html">c.SNPbin</a></td>
<td>Formal class "SNPbin"</td></tr>
<tr><td style="width: 25%;"><a href="virClasses.html">callOrNULL-class</a></td>
<td>Virtual classes for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">cbind.genlight</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="SNPbin.html">cbind.SNPbin</a></td>
<td>Formal class "SNPbin"</td></tr>
<tr><td style="width: 25%;"><a href="virClasses.html">charOrNULL-class</a></td>
<td>Virtual classes for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="doc_C_routines.html">CheckAllSeg</a></td>
<td>Internal C routines</td></tr>
<tr><td style="width: 25%;"><a href="auxil.html">checkType</a></td>
<td>Auxiliary functions for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="chooseCN.html">chooseCN</a></td>
<td>Function to choose a connection network</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">chr</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">chr-method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">chr&lt;-</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">chr&lt;--method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">chromosome</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">chromosome-method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">chromosome&lt;-</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">chromosome&lt;--method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="SNPbin.html">coerce-method</a></td>
<td>Formal class "SNPbin"</td></tr>
<tr><td style="width: 25%;"><a href="as-methods.html">coerce-method</a></td>
<td>Converting genind/genpop objects to other classes</td></tr>
<tr><td style="width: 25%;"><a href="as.SNPbin.html">coerce-method</a></td>
<td>Conversion to class "SNPbin"</td></tr>
<tr><td style="width: 25%;"><a href="as.genlight.html">coerce-method</a></td>
<td>Conversion to class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">coerce-method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="colorplot.html">colorplot</a></td>
<td>Represents a cloud of points with colors</td></tr>
<tr><td style="width: 25%;"><a href="colorplot.html">colorplot.default</a></td>
<td>Represents a cloud of points with colors</td></tr>
<tr><td style="width: 25%;"><a href="spca.html">colorplot.spca</a></td>
<td>Spatial principal component analysis</td></tr>
<tr><td style="width: 25%;"><a href="compoplot.html">compoplot</a></td>
<td>Genotype composition plot</td></tr>
<tr><td style="width: 25%;"><a href="compoplot.html">compoplot.dapc</a></td>
<td>Genotype composition plot</td></tr>
<tr><td style="width: 25%;"><a href="compoplot.html">compoplot.matrix</a></td>
<td>Genotype composition plot</td></tr>
<tr><td style="width: 25%;"><a href="compoplot.html">compoplot.snapclust</a></td>
<td>Genotype composition plot</td></tr>
<tr><td style="width: 25%;"><a href="coords.monmonier.html">coords.monmonier</a></td>
<td>Returns original points in results paths of an object of class 'monmonier'</td></tr>
<tr><td style="width: 25%;"><a href="auxil.html">corner</a></td>
<td>Auxiliary functions for adegenet</td></tr>
</table>

<h2><a name="D">-- D --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="dapc.html">dapc</a></td>
<td>Discriminant Analysis of Principal Components (DAPC)</td></tr>
<tr><td style="width: 25%;"><a href="dapc.html">dapc.data.frame</a></td>
<td>Discriminant Analysis of Principal Components (DAPC)</td></tr>
<tr><td style="width: 25%;"><a href="dapc.html">dapc.dudi</a></td>
<td>Discriminant Analysis of Principal Components (DAPC)</td></tr>
<tr><td style="width: 25%;"><a href="dapc.html">dapc.genind</a></td>
<td>Discriminant Analysis of Principal Components (DAPC)</td></tr>
<tr><td style="width: 25%;"><a href="dapc.html">dapc.genlight</a></td>
<td>Discriminant Analysis of Principal Components (DAPC)</td></tr>
<tr><td style="width: 25%;"><a href="dapc.html">dapc.matrix</a></td>
<td>Discriminant Analysis of Principal Components (DAPC)</td></tr>
<tr><td style="width: 25%;"><a href="dapcIllus.html">dapcIllus</a></td>
<td>Simulated data illustrating the DAPC</td></tr>
<tr><td style="width: 25%;"><a href="auxil.html">deepseasun</a></td>
<td>Auxiliary functions for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="df2genind.html">df2genind</a></td>
<td>Convert a data.frame of allele data to a genind object.</td></tr>
<tr><td style="width: 25%;"><a href="virClasses.html">dfOrNULL-class</a></td>
<td>Virtual classes for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">dim-method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="genpop.html">dist-method</a></td>
<td>adegenet formal class (S4) for allele counts in populations</td></tr>
<tr><td style="width: 25%;"><a href="dist.genpop.html">dist.genpop</a></td>
<td>Genetic distances between populations</td></tr>
<tr><td style="width: 25%;"><a href="sequences.html">DNAbin2genind</a></td>
<td>Importing data from an alignement of sequences to a genind object</td></tr>
</table>

<h2><a name="E">-- E --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="eHGDP.html">eHGDP</a></td>
<td>Extended HGDP-CEPH dataset</td></tr>
<tr><td style="width: 25%;"><a href="export_to_mvmapper.html">export_to_mvmapper</a></td>
<td>Export analysis for mvmapper visualisation</td></tr>
<tr><td style="width: 25%;"><a href="export_to_mvmapper.html">export_to_mvmapper.dapc</a></td>
<td>Export analysis for mvmapper visualisation</td></tr>
<tr><td style="width: 25%;"><a href="export_to_mvmapper.html">export_to_mvmapper.default</a></td>
<td>Export analysis for mvmapper visualisation</td></tr>
<tr><td style="width: 25%;"><a href="export_to_mvmapper.html">export_to_mvmapper.dudi</a></td>
<td>Export analysis for mvmapper visualisation</td></tr>
<tr><td style="width: 25%;"><a href="export_to_mvmapper.html">export_to_mvmapper.spca</a></td>
<td>Export analysis for mvmapper visualisation</td></tr>
<tr><td style="width: 25%;"><a href="read.PLINK.html">extract.PLINKmap</a></td>
<td>Reading PLINK Single Nucleotide Polymorphism data</td></tr>
</table>

<h2><a name="F">-- F --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="auxil.html">fac2col</a></td>
<td>Auxiliary functions for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="virClasses.html">factorOrNULL-class</a></td>
<td>Virtual classes for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="fasta2DNAbin.html">fasta2DNAbin</a></td>
<td>Read large DNA alignments into R</td></tr>
<tr><td style="width: 25%;"><a href="fasta2genlight.html">fasta2genlight</a></td>
<td>Extract Single Nucleotide Polymorphism (SNPs) from alignments</td></tr>
<tr><td style="width: 25%;"><a href="find.clusters.html">find.clusters</a></td>
<td>find.cluster: cluster identification using successive K-means</td></tr>
<tr><td style="width: 25%;"><a href="find.clusters.html">find.clusters.data.frame</a></td>
<td>find.cluster: cluster identification using successive K-means</td></tr>
<tr><td style="width: 25%;"><a href="find.clusters.html">find.clusters.genind</a></td>
<td>find.cluster: cluster identification using successive K-means</td></tr>
<tr><td style="width: 25%;"><a href="find.clusters.html">find.clusters.genlight</a></td>
<td>find.cluster: cluster identification using successive K-means</td></tr>
<tr><td style="width: 25%;"><a href="find.clusters.html">find.clusters.matrix</a></td>
<td>find.cluster: cluster identification using successive K-means</td></tr>
<tr><td style="width: 25%;"><a href="mutations.html">findMutations</a></td>
<td>Identify mutations between DNA sequences</td></tr>
<tr><td style="width: 25%;"><a href="mutations.html">findMutations.DNAbin</a></td>
<td>Identify mutations between DNA sequences</td></tr>
<tr><td style="width: 25%;"><a href="auxil.html">flame</a></td>
<td>Auxiliary functions for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="virClasses.html">formOrNULL-class</a></td>
<td>Virtual classes for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="auxil.html">funky</a></td>
<td>Auxiliary functions for adegenet</td></tr>
</table>

<h2><a name="G">-- G --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="virClasses.html">gen-class</a></td>
<td>Virtual classes for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="gengraph.html">gengraph</a></td>
<td>Genetic transitive graphs</td></tr>
<tr><td style="width: 25%;"><a href="gengraph.html">gengraph.default</a></td>
<td>Genetic transitive graphs</td></tr>
<tr><td style="width: 25%;"><a href="gengraph.html">gengraph.dist</a></td>
<td>Genetic transitive graphs</td></tr>
<tr><td style="width: 25%;"><a href="gengraph.html">gengraph.DNAbin</a></td>
<td>Genetic transitive graphs</td></tr>
<tr><td style="width: 25%;"><a href="gengraph.html">gengraph.genind</a></td>
<td>Genetic transitive graphs</td></tr>
<tr><td style="width: 25%;"><a href="gengraph.html">gengraph.genpop</a></td>
<td>Genetic transitive graphs</td></tr>
<tr><td style="width: 25%;"><a href="gengraph.html">gengraph.matrix</a></td>
<td>Genetic transitive graphs</td></tr>
<tr><td style="width: 25%;"><a href="new.genind.html">genind</a></td>
<td>genind constructor</td></tr>
<tr><td style="width: 25%;"><a href="genind.html">genind-class</a></td>
<td>adegenet formal class (S4) for individual genotypes</td></tr>
<tr><td style="width: 25%;"><a href="genind2df.html">genind2df</a></td>
<td>Convert a genind object to a data.frame.</td></tr>
<tr><td style="width: 25%;"><a href="genind2genpop.html">genind2genpop</a></td>
<td>Conversion from a genind to a genpop object</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">genlight</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">genlight-class</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="new.genpop.html">genpop</a></td>
<td>genpop constructor</td></tr>
<tr><td style="width: 25%;"><a href="genpop.html">genpop-class</a></td>
<td>adegenet formal class (S4) for allele counts in populations</td></tr>
<tr><td style="width: 25%;"><a href="seqTrack.html">get.likelihood</a></td>
<td>SeqTrack algorithm for reconstructing genealogies</td></tr>
<tr><td style="width: 25%;"><a href="seqTrack.html">get.likelihood.seqTrack</a></td>
<td>SeqTrack algorithm for reconstructing genealogies</td></tr>
<tr><td style="width: 25%;"><a href="doc_C_routines.html">GLdotProd</a></td>
<td>Internal C routines</td></tr>
<tr><td style="width: 25%;"><a href="glAux.html">glDotProd</a></td>
<td>Auxiliary functions for genlight objects</td></tr>
<tr><td style="width: 25%;"><a href="glAux.html">glMean</a></td>
<td>Auxiliary functions for genlight objects</td></tr>
<tr><td style="width: 25%;"><a href="glAux.html">glNA</a></td>
<td>Auxiliary functions for genlight objects</td></tr>
<tr><td style="width: 25%;"><a href="spca.rtests.html">global.rtest</a></td>
<td>Global and local tests</td></tr>
<tr><td style="width: 25%;"><a href="glPca.html">glPca</a></td>
<td>Principal Component Analysis for genlight objects</td></tr>
<tr><td style="width: 25%;"><a href="glPlot.html">glPlot</a></td>
<td>Plotting genlight objects</td></tr>
<tr><td style="width: 25%;"><a href="glSim.html">glSim</a></td>
<td>Simulation of simple genlight objects</td></tr>
<tr><td style="width: 25%;"><a href="glAux.html">glSum</a></td>
<td>Auxiliary functions for genlight objects</td></tr>
<tr><td style="width: 25%;"><a href="doc_C_routines.html">GLsumFreq</a></td>
<td>Internal C routines</td></tr>
<tr><td style="width: 25%;"><a href="doc_C_routines.html">GLsumInt</a></td>
<td>Internal C routines</td></tr>
<tr><td style="width: 25%;"><a href="glAux.html">glVar</a></td>
<td>Auxiliary functions for genlight objects</td></tr>
<tr><td style="width: 25%;"><a href="mutations.html">graphMutations</a></td>
<td>Identify mutations between DNA sequences</td></tr>
<tr><td style="width: 25%;"><a href="mutations.html">graphMutations.DNAbin</a></td>
<td>Identify mutations between DNA sequences</td></tr>
<tr><td style="width: 25%;"><a href="auxil.html">greenpal</a></td>
<td>Auxiliary functions for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="auxil.html">greypal</a></td>
<td>Auxiliary functions for adegenet</td></tr>
</table>

<h2><a name="H">-- H --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="H3N2.html">H3N2</a></td>
<td>Seasonal influenza (H3N2) HA segment data</td></tr>
<tr><td style="width: 25%;"><a href="haploGen.html">haploGen</a></td>
<td>Simulation of genealogies of haplotypes</td></tr>
<tr><td style="width: 25%;"><a href="haploGen.html">haploGen-class</a></td>
<td>Simulation of genealogies of haplotypes</td></tr>
<tr><td style="width: 25%;"><a href="hierarchy-methods.html">hier</a></td>
<td>Access and manipulate the population hierarchy for genind or genlight objects.</td></tr>
<tr><td style="width: 25%;"><a href="hierarchy-methods.html">hier-method</a></td>
<td>Access and manipulate the population hierarchy for genind or genlight objects.</td></tr>
<tr><td style="width: 25%;"><a href="hierarchy-methods.html">hier&lt;-</a></td>
<td>Access and manipulate the population hierarchy for genind or genlight objects.</td></tr>
<tr><td style="width: 25%;"><a href="hierarchy-methods.html">hier&lt;--method</a></td>
<td>Access and manipulate the population hierarchy for genind or genlight objects.</td></tr>
<tr><td style="width: 25%;"><a href="Hs.html">Hs</a></td>
<td>Expected heterozygosity (Hs)</td></tr>
<tr><td style="width: 25%;"><a href="Hs.test.html">Hs.test</a></td>
<td>Test differences in expected heterozygosity (Hs)</td></tr>
<tr><td style="width: 25%;"><a href="HWE.html">HWE.test.genind</a></td>
<td>Hardy-Weinberg Equilibrium test for multilocus data</td></tr>
<tr><td style="width: 25%;"><a href="hybridize.html">hybridize</a></td>
<td>Function hybridize takes two genind in inputs and generates hybrids individuals having one parent in both objects.</td></tr>
<tr><td style="width: 25%;"><a href="auxil.html">hybridpal</a></td>
<td>Auxiliary functions for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="hybridtoy.html">hybridtoy</a></td>
<td>Toy hybrid dataset</td></tr>
</table>

<h2><a name="I">-- I --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="import2genind.html">import2genind</a></td>
<td>Importing data from several softwares to a genind object</td></tr>
<tr><td style="width: 25%;"><a href="inbreeding.html">inbreeding</a></td>
<td>Likelihood-based estimation of inbreeding</td></tr>
<tr><td style="width: 25%;"><a href="virClasses.html">indInfo-class</a></td>
<td>Virtual classes for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">indNames</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">indNames-method</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">indNames-method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">indNames&lt;-</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">indNames&lt;--method</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">indNames&lt;--method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="new.genind.html">initialize,genind-methods</a></td>
<td>genind constructor</td></tr>
<tr><td style="width: 25%;"><a href="new.genpop.html">initialize,genpop-methods</a></td>
<td>genpop constructor</td></tr>
<tr><td style="width: 25%;"><a href="SNPbin.html">initialize-method</a></td>
<td>Formal class "SNPbin"</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">initialize-method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="new.genind.html">initialize-method</a></td>
<td>genind constructor</td></tr>
<tr><td style="width: 25%;"><a href="new.genpop.html">initialize-method</a></td>
<td>genpop constructor</td></tr>
<tr><td style="width: 25%;"><a href="virClasses.html">intOrNULL-class</a></td>
<td>Virtual classes for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="virClasses.html">intOrNum-class</a></td>
<td>Virtual classes for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="genind.html">is.genind</a></td>
<td>adegenet formal class (S4) for individual genotypes</td></tr>
<tr><td style="width: 25%;"><a href="genpop.html">is.genpop</a></td>
<td>adegenet formal class (S4) for allele counts in populations</td></tr>
<tr><td style="width: 25%;"><a href="isPoly.html">isPoly</a></td>
<td>Assess polymorphism in genind/genpop objects</td></tr>
<tr><td style="width: 25%;"><a href="isPoly.html">isPoly-method</a></td>
<td>Assess polymorphism in genind/genpop objects</td></tr>
<tr><td style="width: 25%;"><a href="isPoly.html">isPoly-methods</a></td>
<td>Assess polymorphism in genind/genpop objects</td></tr>
</table>

<h2><a name="K">-- K --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="KIC.html">KIC</a></td>
<td>Compute Akaike Information Criterion for small samples (AICc) for snapclust</td></tr>
<tr><td style="width: 25%;"><a href="KIC.html">KIC.snapclust</a></td>
<td>Compute Akaike Information Criterion for small samples (AICc) for snapclust</td></tr>
<tr><td style="width: 25%;"><a href="as-methods.html">ktab-class</a></td>
<td>Converting genind/genpop objects to other classes</td></tr>
</table>

<h2><a name="L">-- L --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="haploGen.html">labels.haploGen</a></td>
<td>Simulation of genealogies of haplotypes</td></tr>
<tr><td style="width: 25%;"><a href="auxil.html">lightseasun</a></td>
<td>Auxiliary functions for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="virClasses.html">listOrNULL-class</a></td>
<td>Virtual classes for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="loadingplot.html">loadingplot</a></td>
<td>Represents a cloud of points with colors</td></tr>
<tr><td style="width: 25%;"><a href="loadingplot.html">loadingplot.default</a></td>
<td>Represents a cloud of points with colors</td></tr>
<tr><td style="width: 25%;"><a href="glPca.html">loadingplot.glPca</a></td>
<td>Principal Component Analysis for genlight objects</td></tr>
<tr><td style="width: 25%;"><a href="spca.rtests.html">local.rtest</a></td>
<td>Global and local tests</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">locFac</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">locFac-method</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">locNames</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">locNames-method</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">locNames-method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">locNames&lt;-</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">locNames&lt;--method</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">locNames&lt;--method</a></td>
<td>Formal class "genlight"</td></tr>
</table>

<h2><a name="M">-- M --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="makefreq.html">makefreq</a></td>
<td>Compute allelic frequencies</td></tr>
<tr><td style="width: 25%;"><a href="makefreq.html">makefreq,genind-methods</a></td>
<td>Compute allelic frequencies</td></tr>
<tr><td style="width: 25%;"><a href="makefreq.html">makefreq,genpop-methods</a></td>
<td>Compute allelic frequencies</td></tr>
<tr><td style="width: 25%;"><a href="makefreq.html">makefreq-method</a></td>
<td>Compute allelic frequencies</td></tr>
<tr><td style="width: 25%;"><a href="makefreq.html">makefreq.genind</a></td>
<td>Compute allelic frequencies</td></tr>
<tr><td style="width: 25%;"><a href="makefreq.html">makefreq.genpop</a></td>
<td>Compute allelic frequencies</td></tr>
<tr><td style="width: 25%;"><a href="microbov.html">microbov</a></td>
<td>Microsatellites genotypes of 15 cattle breeds</td></tr>
<tr><td style="width: 25%;"><a href="minorAllele.html">minorAllele</a></td>
<td>Compute minor allele frequency</td></tr>
<tr><td style="width: 25%;"><a href="monmonier.html">monmonier</a></td>
<td>Boundary detection using Monmonier algorithm</td></tr>
</table>

<h2><a name="N">-- N --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="genlight.html">NA.posi</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="SNPbin.html">NA.posi-method</a></td>
<td>Formal class "SNPbin"</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">NA.posi-method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">nAll</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">nAll-method</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="SNPbin.html">names-method</a></td>
<td>Formal class "SNPbin"</td></tr>
<tr><td style="width: 25%;"><a href="genind.html">names-method</a></td>
<td>adegenet formal class (S4) for individual genotypes</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">names-method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="genpop.html">names-method</a></td>
<td>adegenet formal class (S4) for allele counts in populations</td></tr>
<tr><td style="width: 25%;"><a href="strata-methods.html">nameStrata</a></td>
<td>Access and manipulate the population strata for genind or genlight objects.</td></tr>
<tr><td style="width: 25%;"><a href="strata-methods.html">nameStrata-method</a></td>
<td>Access and manipulate the population strata for genind or genlight objects.</td></tr>
<tr><td style="width: 25%;"><a href="strata-methods.html">nameStrata&lt;-</a></td>
<td>Access and manipulate the population strata for genind or genlight objects.</td></tr>
<tr><td style="width: 25%;"><a href="strata-methods.html">nameStrata&lt;--method</a></td>
<td>Access and manipulate the population strata for genind or genlight objects.</td></tr>
<tr><td style="width: 25%;"><a href="nancycats.html">nancycats</a></td>
<td>Microsatellites genotypes of 237 cats from 17 colonies of Nancy (France)</td></tr>
<tr><td style="width: 25%;"><a href="doc_C_routines.html">nb_shared_all</a></td>
<td>Internal C routines</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">nInd</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">nInd-method</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">nInd-method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">nLoc</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="SNPbin.html">nLoc-method</a></td>
<td>Formal class "SNPbin"</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">nLoc-method</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">nLoc-method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">nPop</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">nPop-method</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">nPop-method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="auxil.html">num2col</a></td>
<td>Auxiliary functions for adegenet</td></tr>
</table>

<h2><a name="O">-- O --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="old2new.html">old2new</a></td>
<td>Convert objects with obsolete classes into new objects</td></tr>
<tr><td style="width: 25%;"><a href="old2new.html">old2new_genind</a></td>
<td>Convert objects with obsolete classes into new objects</td></tr>
<tr><td style="width: 25%;"><a href="old2new.html">old2new_genlight</a></td>
<td>Convert objects with obsolete classes into new objects</td></tr>
<tr><td style="width: 25%;"><a href="old2new.html">old2new_genpop</a></td>
<td>Convert objects with obsolete classes into new objects</td></tr>
<tr><td style="width: 25%;"><a href="ascore.html">optim.a.score</a></td>
<td>Compute and optimize a-score for Discriminant Analysis of Principal Components (DAPC)</td></tr>
<tr><td style="width: 25%;"><a href="monmonier.html">optimize.monmonier</a></td>
<td>Boundary detection using Monmonier algorithm</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">other</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">other-method</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">other-method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">other&lt;-</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">other&lt;--method</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">other&lt;--method</a></td>
<td>Formal class "genlight"</td></tr>
</table>

<h2><a name="P">-- P --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="pairDist.html">pairDist</a></td>
<td>Pairwise distance plots</td></tr>
<tr><td style="width: 25%;"><a href="pairDist.html">pairDist.default</a></td>
<td>Pairwise distance plots</td></tr>
<tr><td style="width: 25%;"><a href="pairDist.html">pairDistPlot</a></td>
<td>Pairwise distance plots</td></tr>
<tr><td style="width: 25%;"><a href="pairDist.html">pairDistPlot.default</a></td>
<td>Pairwise distance plots</td></tr>
<tr><td style="width: 25%;"><a href="pairDist.html">pairDistPlot.dist</a></td>
<td>Pairwise distance plots</td></tr>
<tr><td style="width: 25%;"><a href="pairDist.html">pairDistPlot.DNAbin</a></td>
<td>Pairwise distance plots</td></tr>
<tr><td style="width: 25%;"><a href="pairDist.html">pairDistPlot.genind</a></td>
<td>Pairwise distance plots</td></tr>
<tr><td style="width: 25%;"><a href="pairDist.html">pairDistPlot.matrix</a></td>
<td>Pairwise distance plots</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">ploidy</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="SNPbin.html">ploidy-method</a></td>
<td>Formal class "SNPbin"</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">ploidy-method</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">ploidy-method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">ploidy&lt;-</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="SNPbin.html">ploidy&lt;--method</a></td>
<td>Formal class "SNPbin"</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">ploidy&lt;--method</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">ploidy&lt;--method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="glPlot.html">plot-method</a></td>
<td>Plotting genlight objects</td></tr>
<tr><td style="width: 25%;"><a href="glPlot.html">plot.genlight</a></td>
<td>Plotting genlight objects</td></tr>
<tr><td style="width: 25%;"><a href="haploGen.html">plot.haploGen</a></td>
<td>Simulation of genealogies of haplotypes</td></tr>
<tr><td style="width: 25%;"><a href="monmonier.html">plot.monmonier</a></td>
<td>Boundary detection using Monmonier algorithm</td></tr>
<tr><td style="width: 25%;"><a href="seqTrack.html">plot.seqTrack</a></td>
<td>SeqTrack algorithm for reconstructing genealogies</td></tr>
<tr><td style="width: 25%;"><a href="spca.html">plot.spca</a></td>
<td>Spatial principal component analysis</td></tr>
<tr><td style="width: 25%;"><a href="haploGen.html">plotHaploGen</a></td>
<td>Simulation of genealogies of haplotypes</td></tr>
<tr><td style="width: 25%;"><a href="seqTrack.html">plotSeqTrack</a></td>
<td>SeqTrack algorithm for reconstructing genealogies</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">pop</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">pop-method</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">pop-method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">pop&lt;-</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">pop&lt;--method</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">pop&lt;--method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="virClasses.html">popInfo-class</a></td>
<td>Virtual classes for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">popNames</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">popNames-method</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">popNames-method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">popNames&lt;-</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">popNames&lt;--method</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">popNames&lt;--method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">position</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">position-method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">position&lt;-</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">position&lt;--method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="dapc.html">predict.dapc</a></td>
<td>Discriminant Analysis of Principal Components (DAPC)</td></tr>
<tr><td style="width: 25%;"><a href="genind.html">print-method</a></td>
<td>adegenet formal class (S4) for individual genotypes</td></tr>
<tr><td style="width: 25%;"><a href="genpop.html">print-method</a></td>
<td>adegenet formal class (S4) for allele counts in populations</td></tr>
<tr><td style="width: 25%;"><a href="dapc.html">print.dapc</a></td>
<td>Discriminant Analysis of Principal Components (DAPC)</td></tr>
<tr><td style="width: 25%;"><a href="genind.html">print.genindSummary</a></td>
<td>adegenet formal class (S4) for individual genotypes</td></tr>
<tr><td style="width: 25%;"><a href="genpop.html">print.genpopSummary</a></td>
<td>adegenet formal class (S4) for allele counts in populations</td></tr>
<tr><td style="width: 25%;"><a href="glPca.html">print.glPca</a></td>
<td>Principal Component Analysis for genlight objects</td></tr>
<tr><td style="width: 25%;"><a href="haploGen.html">print.haploGen</a></td>
<td>Simulation of genealogies of haplotypes</td></tr>
<tr><td style="width: 25%;"><a href="monmonier.html">print.monmonier</a></td>
<td>Boundary detection using Monmonier algorithm</td></tr>
<tr><td style="width: 25%;"><a href="spca.html">print.spca</a></td>
<td>Spatial principal component analysis</td></tr>
<tr><td style="width: 25%;"><a href="propShared.html">propShared</a></td>
<td>Compute proportion of shared alleles</td></tr>
<tr><td style="width: 25%;"><a href="propTyped.html">propTyped</a></td>
<td>Compute the proportion of typed elements</td></tr>
<tr><td style="width: 25%;"><a href="propTyped.html">propTyped-method</a></td>
<td>Compute the proportion of typed elements</td></tr>
<tr><td style="width: 25%;"><a href="propTyped.html">propTyped-methods</a></td>
<td>Compute the proportion of typed elements</td></tr>
</table>

<h2><a name="R">-- R --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="genlight.html">rbind.genlight</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="read.fstat.html">read.fstat</a></td>
<td>Reading data from Fstat</td></tr>
<tr><td style="width: 25%;"><a href="read.genepop.html">read.genepop</a></td>
<td>Reading data from Genepop</td></tr>
<tr><td style="width: 25%;"><a href="read.genetix.html">read.genetix</a></td>
<td>Reading data from GENETIX</td></tr>
<tr><td style="width: 25%;"><a href="read.PLINK.html">read.PLINK</a></td>
<td>Reading PLINK Single Nucleotide Polymorphism data</td></tr>
<tr><td style="width: 25%;"><a href="read.PLINK.html">read.plink</a></td>
<td>Reading PLINK Single Nucleotide Polymorphism data</td></tr>
<tr><td style="width: 25%;"><a href="read.snp.html">read.snp</a></td>
<td>Reading Single Nucleotide Polymorphism data</td></tr>
<tr><td style="width: 25%;"><a href="read.structure.html">read.structure</a></td>
<td>Reading data from STRUCTURE</td></tr>
<tr><td style="width: 25%;"><a href="auxil.html">redpal</a></td>
<td>Auxiliary functions for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="repool.html">repool</a></td>
<td>Pool several genotypes into a single dataset</td></tr>
<tr><td style="width: 25%;"><a href="rupica.html">rupica</a></td>
<td>Microsatellites genotypes of 335 chamois (Rupicapra rupicapra) from the Bauges mountains (France)</td></tr>
</table>

<h2><a name="S">-- S --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="haploGen.html">sample.haploGen</a></td>
<td>Simulation of genealogies of haplotypes</td></tr>
<tr><td style="width: 25%;"><a href="scaleGen.html">scaleGen</a></td>
<td>Compute scaled allele frequencies</td></tr>
<tr><td style="width: 25%;"><a href="scaleGen.html">scaleGen-method</a></td>
<td>Compute scaled allele frequencies</td></tr>
<tr><td style="width: 25%;"><a href="scaleGen.html">scaleGen-methods</a></td>
<td>Compute scaled allele frequencies</td></tr>
<tr><td style="width: 25%;"><a href="dapcGraphics.html">scatter.dapc</a></td>
<td>Graphics for Discriminant Analysis of Principal Components (DAPC)</td></tr>
<tr><td style="width: 25%;"><a href="glPca.html">scatter.glPca</a></td>
<td>Principal Component Analysis for genlight objects</td></tr>
<tr><td style="width: 25%;"><a href="spca.html">screeplot.spca</a></td>
<td>Spatial principal component analysis</td></tr>
<tr><td style="width: 25%;"><a href="auxil.html">seasun</a></td>
<td>Auxiliary functions for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="selpopsize.html">selPopSize</a></td>
<td>Select genotypes of well-represented populations</td></tr>
<tr><td style="width: 25%;"><a href="selpopsize.html">selPopSize-method</a></td>
<td>Select genotypes of well-represented populations</td></tr>
<tr><td style="width: 25%;"><a href="selpopsize.html">selPopSize-methods</a></td>
<td>Select genotypes of well-represented populations</td></tr>
<tr><td style="width: 25%;"><a href="seploc.html">seploc</a></td>
<td>Separate data per locus</td></tr>
<tr><td style="width: 25%;"><a href="seploc.html">seploc-method</a></td>
<td>Separate data per locus</td></tr>
<tr><td style="width: 25%;"><a href="seploc.html">seploc-methods</a></td>
<td>Separate data per locus</td></tr>
<tr><td style="width: 25%;"><a href="seppop.html">seppop</a></td>
<td>Separate genotypes per population</td></tr>
<tr><td style="width: 25%;"><a href="seppop.html">seppop-method</a></td>
<td>Separate genotypes per population</td></tr>
<tr><td style="width: 25%;"><a href="seppop.html">seppop-methods</a></td>
<td>Separate genotypes per population</td></tr>
<tr><td style="width: 25%;"><a href="seqTrack.html">seqTrack</a></td>
<td>SeqTrack algorithm for reconstructing genealogies</td></tr>
<tr><td style="width: 25%;"><a href="seqTrack.html">seqTrack-class</a></td>
<td>SeqTrack algorithm for reconstructing genealogies</td></tr>
<tr><td style="width: 25%;"><a href="seqTrack.html">seqTrack.default</a></td>
<td>SeqTrack algorithm for reconstructing genealogies</td></tr>
<tr><td style="width: 25%;"><a href="haploGen.html">seqTrack.haploGen</a></td>
<td>Simulation of genealogies of haplotypes</td></tr>
<tr><td style="width: 25%;"><a href="seqTrack.html">seqTrack.matrix</a></td>
<td>SeqTrack algorithm for reconstructing genealogies</td></tr>
<tr><td style="width: 25%;"><a href="population-methods.html">setPop</a></td>
<td>Manipulate the population factor of genind objects.</td></tr>
<tr><td style="width: 25%;"><a href="population-methods.html">setPop-method</a></td>
<td>Manipulate the population factor of genind objects.</td></tr>
<tr><td style="width: 25%;"><a href="population-methods.html">setPop&lt;-</a></td>
<td>Manipulate the population factor of genind objects.</td></tr>
<tr><td style="width: 25%;"><a href="population-methods.html">setPop&lt;--method</a></td>
<td>Manipulate the population factor of genind objects.</td></tr>
<tr><td style="width: 25%;"><a href="SNPbin.html">show-method</a></td>
<td>Formal class "SNPbin"</td></tr>
<tr><td style="width: 25%;"><a href="genind.html">show-method</a></td>
<td>adegenet formal class (S4) for individual genotypes</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">show-method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="genpop.html">show-method</a></td>
<td>adegenet formal class (S4) for allele counts in populations</td></tr>
<tr><td style="width: 25%;"><a href="showmekittens.html">showmekittens</a></td>
<td>When you need a break...</td></tr>
<tr><td style="width: 25%;"><a href="sim2pop.html">sim2pop</a></td>
<td>Simulated genotypes of two georeferenced populations</td></tr>
<tr><td style="width: 25%;"><a href="snapclust.html">snapclust</a></td>
<td>Maximum-likelihood genetic clustering using EM algorithm</td></tr>
<tr><td style="width: 25%;"><a href="snapclust.choose.k.html">snapclust.choose.k</a></td>
<td>Choose the number of clusters for snapclust using BIC</td></tr>
<tr><td style="width: 25%;"><a href="SNPbin.html">SNPbin</a></td>
<td>Formal class "SNPbin"</td></tr>
<tr><td style="width: 25%;"><a href="SNPbin.html">SNPbin-class</a></td>
<td>Formal class "SNPbin"</td></tr>
<tr><td style="width: 25%;"><a href="snpposi.html">snpposi.plot</a></td>
<td>Analyse the position of polymorphic sites</td></tr>
<tr><td style="width: 25%;"><a href="snpposi.html">snpposi.plot.DNAbin</a></td>
<td>Analyse the position of polymorphic sites</td></tr>
<tr><td style="width: 25%;"><a href="snpposi.html">snpposi.plot.integer</a></td>
<td>Analyse the position of polymorphic sites</td></tr>
<tr><td style="width: 25%;"><a href="snpposi.html">snpposi.plot.numeric</a></td>
<td>Analyse the position of polymorphic sites</td></tr>
<tr><td style="width: 25%;"><a href="snpposi.html">snpposi.test</a></td>
<td>Analyse the position of polymorphic sites</td></tr>
<tr><td style="width: 25%;"><a href="snpposi.html">snpposi.test.DNAbin</a></td>
<td>Analyse the position of polymorphic sites</td></tr>
<tr><td style="width: 25%;"><a href="snpposi.html">snpposi.test.integer</a></td>
<td>Analyse the position of polymorphic sites</td></tr>
<tr><td style="width: 25%;"><a href="snpposi.html">snpposi.test.numeric</a></td>
<td>Analyse the position of polymorphic sites</td></tr>
<tr><td style="width: 25%;"><a href="snpzip.html">snpzip</a></td>
<td>Identification of structural SNPs</td></tr>
<tr><td style="width: 25%;"><a href="spca.html">spca</a></td>
<td>Spatial principal component analysis</td></tr>
<tr><td style="width: 25%;"><a href="spca.html">spca.data.frame</a></td>
<td>Spatial principal component analysis</td></tr>
<tr><td style="width: 25%;"><a href="spca.html">spca.default</a></td>
<td>Spatial principal component analysis</td></tr>
<tr><td style="width: 25%;"><a href="spca.html">spca.genind</a></td>
<td>Spatial principal component analysis</td></tr>
<tr><td style="width: 25%;"><a href="spca.html">spca.genpop</a></td>
<td>Spatial principal component analysis</td></tr>
<tr><td style="width: 25%;"><a href="spca.html">spca.matrix</a></td>
<td>Spatial principal component analysis</td></tr>
<tr><td style="width: 25%;"><a href="spcaIllus.html">spcaIllus</a></td>
<td>Simulated data illustrating the sPCA</td></tr>
<tr><td style="width: 25%;"><a href="spca_randtest.html">spca_randtest</a></td>
<td>Monte Carlo test for sPCA</td></tr>
<tr><td style="width: 25%;"><a href="auxil.html">spectral</a></td>
<td>Auxiliary functions for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="strata-methods.html">splitStrata</a></td>
<td>Access and manipulate the population strata for genind or genlight objects.</td></tr>
<tr><td style="width: 25%;"><a href="strata-methods.html">splitStrata-method</a></td>
<td>Access and manipulate the population strata for genind or genlight objects.</td></tr>
<tr><td style="width: 25%;"><a href="strata-methods.html">splitStrata&lt;-</a></td>
<td>Access and manipulate the population strata for genind or genlight objects.</td></tr>
<tr><td style="width: 25%;"><a href="strata-methods.html">splitStrata&lt;--method</a></td>
<td>Access and manipulate the population strata for genind or genlight objects.</td></tr>
<tr><td style="width: 25%;"><a href="strata-methods.html">strata</a></td>
<td>Access and manipulate the population strata for genind or genlight objects.</td></tr>
<tr><td style="width: 25%;"><a href="strata-methods.html">strata-method</a></td>
<td>Access and manipulate the population strata for genind or genlight objects.</td></tr>
<tr><td style="width: 25%;"><a href="strata-methods.html">strata&lt;-</a></td>
<td>Access and manipulate the population strata for genind or genlight objects.</td></tr>
<tr><td style="width: 25%;"><a href="strata-methods.html">strata&lt;--method</a></td>
<td>Access and manipulate the population strata for genind or genlight objects.</td></tr>
<tr><td style="width: 25%;"><a href="genind.html">summary-method</a></td>
<td>adegenet formal class (S4) for individual genotypes</td></tr>
<tr><td style="width: 25%;"><a href="genpop.html">summary-method</a></td>
<td>adegenet formal class (S4) for allele counts in populations</td></tr>
<tr><td style="width: 25%;"><a href="dapc.html">summary.dapc</a></td>
<td>Discriminant Analysis of Principal Components (DAPC)</td></tr>
<tr><td style="width: 25%;"><a href="spca.html">summary.spca</a></td>
<td>Spatial principal component analysis</td></tr>
<tr><td style="width: 25%;"><a href="swallowtails.html">swallowtails</a></td>
<td>Microsatellites genotypes of 781 swallowtail butterflies from 40 populations in Alberta and British Columbia, Canada</td></tr>
</table>

<h2><a name="T">-- T --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="tab.html">tab</a></td>
<td>Access allele counts or frequencies</td></tr>
<tr><td style="width: 25%;"><a href="tab.html">tab,genind-methods</a></td>
<td>Access allele counts or frequencies</td></tr>
<tr><td style="width: 25%;"><a href="tab.html">tab,genpop-methods</a></td>
<td>Access allele counts or frequencies</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">tab-method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="tab.html">tab-method</a></td>
<td>Access allele counts or frequencies</td></tr>
<tr><td style="width: 25%;"><a href="tab.html">tab.genind</a></td>
<td>Access allele counts or frequencies</td></tr>
<tr><td style="width: 25%;"><a href="tab.html">tab.genpop</a></td>
<td>Access allele counts or frequencies</td></tr>
<tr><td style="width: 25%;"><a href="auxil.html">transp</a></td>
<td>Auxiliary functions for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="truenames.html">truenames</a></td>
<td>Restore true labels of an object</td></tr>
<tr><td style="width: 25%;"><a href="truenames.html">truenames-method</a></td>
<td>Restore true labels of an object</td></tr>
<tr><td style="width: 25%;"><a href="truenames.html">truenames-methods</a></td>
<td>Restore true labels of an object</td></tr>
</table>

<h2><a name="U">-- U --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="H3N2.html">USflu</a></td>
<td>Seasonal influenza (H3N2) HA segment data</td></tr>
<tr><td style="width: 25%;"><a href="H3N2.html">usflu</a></td>
<td>Seasonal influenza (H3N2) HA segment data</td></tr>
<tr><td style="width: 25%;"><a href="H3N2.html">USflu.fasta</a></td>
<td>Seasonal influenza (H3N2) HA segment data</td></tr>
<tr><td style="width: 25%;"><a href="H3N2.html">usflu.fasta</a></td>
<td>Seasonal influenza (H3N2) HA segment data</td></tr>
</table>

<h2><a name="V">-- V --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="auxil.html">virid</a></td>
<td>Auxiliary functions for adegenet</td></tr>
</table>

<h2><a name="W">-- W --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="auxil.html">wasp</a></td>
<td>Auxiliary functions for adegenet</td></tr>
</table>

<h2><a name="X">-- X --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="xvalDapc.html">xvalDapc</a></td>
<td>Cross-validation for Discriminant Analysis of Principal Components (DAPC)</td></tr>
<tr><td style="width: 25%;"><a href="xvalDapc.html">xvalDapc.data.frame</a></td>
<td>Cross-validation for Discriminant Analysis of Principal Components (DAPC)</td></tr>
<tr><td style="width: 25%;"><a href="xvalDapc.html">xvalDapc.default</a></td>
<td>Cross-validation for Discriminant Analysis of Principal Components (DAPC)</td></tr>
<tr><td style="width: 25%;"><a href="xvalDapc.html">xvalDapc.genind</a></td>
<td>Cross-validation for Discriminant Analysis of Principal Components (DAPC)</td></tr>
<tr><td style="width: 25%;"><a href="xvalDapc.html">xvalDapc.genlight</a></td>
<td>Cross-validation for Discriminant Analysis of Principal Components (DAPC)</td></tr>
<tr><td style="width: 25%;"><a href="xvalDapc.html">xvalDapc.matrix</a></td>
<td>Cross-validation for Discriminant Analysis of Principal Components (DAPC)</td></tr>
</table>

<h2><a name="misc">-- misc --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="SNPbin.html">$-method</a></td>
<td>Formal class "SNPbin"</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">$-method</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">$-method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="SNPbin.html">$&lt;--method</a></td>
<td>Formal class "SNPbin"</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">$&lt;--method</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">$&lt;--method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="find.clusters.html">.find.sub.clusters</a></td>
<td>find.cluster: cluster identification using successive K-means</td></tr>
<tr><td style="width: 25%;"><a href="auxil.html">.genlab</a></td>
<td>Auxiliary functions for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="doc_C_routines.html">.internal_C_routines</a></td>
<td>Internal C routines</td></tr>
<tr><td style="width: 25%;"><a href="auxil.html">.readExt</a></td>
<td>Auxiliary functions for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="auxil.html">.rmspaces</a></td>
<td>Auxiliary functions for adegenet</td></tr>
<tr><td style="width: 25%;"><a href="genind.html">.valid.genind</a></td>
<td>adegenet formal class (S4) for individual genotypes</td></tr>
<tr><td style="width: 25%;"><a href="SNPbin.html">[-method</a></td>
<td>Formal class "SNPbin"</td></tr>
<tr><td style="width: 25%;"><a href="accessors.html">[-method</a></td>
<td>Accessors for adegenet objects</td></tr>
<tr><td style="width: 25%;"><a href="genlight.html">[-method</a></td>
<td>Formal class "genlight"</td></tr>
<tr><td style="width: 25%;"><a href="haploGen.html">[.haploGen</a></td>
<td>Simulation of genealogies of haplotypes</td></tr>
</table>
</body></html>
