<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head><title>R: Phylogenetic Tools for Comparative Biology (and Other Things)</title>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<link rel="stylesheet" type="text/css" href="R.css" />
</head><body>
<h1> Phylogenetic Tools for Comparative Biology (and Other Things)
<img class="toplogo" src="../../../doc/html/Rlogo.svg" alt="[R logo]" />
</h1>
<hr/>
<div style="text-align: center;">
<a href="../../../doc/html/packages.html"><img class="arrow" src="../../../doc/html/left.jpg" alt="[Up]" /></a>
<a href="../../../doc/html/index.html"><img class="arrow" src="../../../doc/html/up.jpg" alt="[Top]" /></a>
</div><h2>Documentation for package &lsquo;phytools&rsquo; version 0.7-47</h2>

<ul><li><a href="../DESCRIPTION">DESCRIPTION file</a>.</li>
</ul>

<h2>Help Pages</h2>


<p style="text-align: center;">
<a href="# "> </a>
<a href="#A">A</a>
<a href="#B">B</a>
<a href="#C">C</a>
<a href="#D">D</a>
<a href="#E">E</a>
<a href="#F">F</a>
<a href="#G">G</a>
<a href="#L">L</a>
<a href="#M">M</a>
<a href="#N">N</a>
<a href="#O">O</a>
<a href="#P">P</a>
<a href="#R">R</a>
<a href="#S">S</a>
<a href="#T">T</a>
<a href="#U">U</a>
<a href="#V">V</a>
<a href="#W">W</a>
</p>

<table width="100%">
<tr><td style="width: 25%;"><a href="phytools-package.html">phytools-package</a></td>
<td>phytools: Phylogenetic Tools for comparative biology (and other things)</td></tr>
</table>

<h2><a name="A">-- A --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="add.arrow.html">add.arrow</a></td>
<td>Add an arrow pointing to a tip or node on the tree</td></tr>
<tr><td style="width: 25%;"><a href="add.color.bar.html">add.color.bar</a></td>
<td>Add color bar to a plot</td></tr>
<tr><td style="width: 25%;"><a href="add.everywhere.html">add.everywhere</a></td>
<td>Add tip to all edges in a tree</td></tr>
<tr><td style="width: 25%;"><a href="add.random.html">add.random</a></td>
<td>Add tips at random to the tree</td></tr>
<tr><td style="width: 25%;"><a href="add.simmap.legend.html">add.simmap.legend</a></td>
<td>Add legend to stochastically mapped tree</td></tr>
<tr><td style="width: 25%;"><a href="add.species.to.genus.html">add.species.to.genus</a></td>
<td>Add species to genus on a phylogeny or bind simulated species subtrees to a backbone genus tree</td></tr>
<tr><td style="width: 25%;"><a href="aic.w.html">aic.w</a></td>
<td>Computes AIC weights</td></tr>
<tr><td style="width: 25%;"><a href="allFurcTrees.html">allFurcTrees</a></td>
<td>Generate all bi- and multifurcating unrooted trees</td></tr>
<tr><td style="width: 25%;"><a href="rotateNodes.html">allRotations</a></td>
<td>Rotates a node or set of nodes in a phylogenetic tree</td></tr>
<tr><td style="width: 25%;"><a href="anc.Bayes.html">anc.Bayes</a></td>
<td>Bayesian ancestral character estimation</td></tr>
<tr><td style="width: 25%;"><a href="anc.ML.html">anc.ML</a></td>
<td>Ancestral character estimation using likelihood</td></tr>
<tr><td style="width: 25%;"><a href="anc.trend.html">anc.trend</a></td>
<td>Ancestral character estimation with a trend</td></tr>
<tr><td style="width: 25%;"><a href="ancThresh.html">ancThresh</a></td>
<td>Ancestral character estimation under the threshold model using Bayesian MCMC</td></tr>
<tr><td style="width: 25%;"><a href="anoletree.html">anole.data</a></td>
<td>Phylogenetic datasets</td></tr>
<tr><td style="width: 25%;"><a href="anoletree.html">anoletree</a></td>
<td>Phylogenetic datasets</td></tr>
<tr><td style="width: 25%;"><a href="applyBranchLengths.html">applyBranchLengths</a></td>
<td>Applies the branch lengths of a reference tree to a target</td></tr>
<tr><td style="width: 25%;"><a href="cladelabels.html">arc.cladelabels</a></td>
<td>Add labels to subtrees of a plotted phylogeny</td></tr>
<tr><td style="width: 25%;"><a href="as.multiPhylo.html">as.multiPhylo</a></td>
<td>Conversion to object of class '"multiPhylo"'</td></tr>
<tr><td style="width: 25%;"><a href="as.multiPhylo.html">as.multiPhylo.multiSimmap</a></td>
<td>Conversion to object of class '"multiPhylo"'</td></tr>
<tr><td style="width: 25%;"><a href="as.multiPhylo.html">as.multiPhylo.phylo</a></td>
<td>Conversion to object of class '"multiPhylo"'</td></tr>
<tr><td style="width: 25%;"><a href="as.Qmatrix.html">as.Qmatrix</a></td>
<td>Convert a fitted M_k_ model to a Q-matrix</td></tr>
<tr><td style="width: 25%;"><a href="as.Qmatrix.html">as.Qmatrix.fitMk</a></td>
<td>Convert a fitted M_k_ model to a Q-matrix</td></tr>
<tr><td style="width: 25%;"><a href="ave.rates.html">ave.rates</a></td>
<td>Average the posterior rates</td></tr>
<tr><td style="width: 25%;"><a href="averageTree.html">averageTree</a></td>
<td>Compute an average tree from a set of trees and related operations</td></tr>
</table>

<h2><a name="B">-- B --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="phylo.toBackbone.html">backbone.toPhylo</a></td>
<td>Converts tree to backbone or vice versa</td></tr>
<tr><td style="width: 25%;"><a href="bd.html">bd</a></td>
<td>Convert object of class "birthdeath" to raw birth &amp; death rates</td></tr>
<tr><td style="width: 25%;"><a href="bind.tip.html">bind.tip</a></td>
<td>Attaches a new tip to a tree</td></tr>
<tr><td style="width: 25%;"><a href="bind.tree.simmap.html">bind.tree.simmap</a></td>
<td>Attaches a new tip to a tree</td></tr>
<tr><td style="width: 25%;"><a href="phyl.pca.html">biplot.phyl.pca</a></td>
<td>Phylogenetic principal components analysis</td></tr>
<tr><td style="width: 25%;"><a href="bmPlot.html">bmPlot</a></td>
<td>Simulates and visualizes discrete-time Brownian evolution on a phylogeny</td></tr>
<tr><td style="width: 25%;"><a href="branching.diffusion.html">branching.diffusion</a></td>
<td>Animation of branching random diffusion</td></tr>
<tr><td style="width: 25%;"><a href="brownie.lite.html">brownie.lite</a></td>
<td>Likelihood test for rate variation in a continuous trait</td></tr>
<tr><td style="width: 25%;"><a href="brownieREML.html">brownieREML</a></td>
<td>REML version of brownie.lite</td></tr>
</table>

<h2><a name="C">-- C --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="cladelabels.html">cladelabels</a></td>
<td>Add labels to subtrees of a plotted phylogeny</td></tr>
<tr><td style="width: 25%;"><a href="phyl.RMA.html">coef.phyl.RMA</a></td>
<td>Phylogenetic reduced major axis (RMA) regression</td></tr>
<tr><td style="width: 25%;"><a href="collapse.to.star.html">collapse.to.star</a></td>
<td>Collapse a subtree to a star phylogeny</td></tr>
<tr><td style="width: 25%;"><a href="collapseTree.html">collapseTree</a></td>
<td>Interactive tree visualizer</td></tr>
<tr><td style="width: 25%;"><a href="compare.chronograms.html">compare.chronograms</a></td>
<td>Compares two chronograms with precisely matching nodes in a visual manner</td></tr>
<tr><td style="width: 25%;"><a href="mrp.supertree.html">compute.mr</a></td>
<td>Matrix representation parsimony supertree estimation</td></tr>
<tr><td style="width: 25%;"><a href="consensus.edges.html">consensus.edges</a></td>
<td>Compute consensus edges for a tree under some criterion</td></tr>
<tr><td style="width: 25%;"><a href="contMap.html">contMap</a></td>
<td>Map continuous trait evolution on the tree</td></tr>
<tr><td style="width: 25%;"><a href="cophylo.html">cophylo</a></td>
<td>Creates a co-phylogenetic plot</td></tr>
<tr><td style="width: 25%;"><a href="cospeciation.html">cospeciation</a></td>
<td>Conducts a statistical test of cospeciation between two trees</td></tr>
<tr><td style="width: 25%;"><a href="countSimmap.html">countSimmap</a></td>
<td>Counts the number of character changes on a object of class "simmap" or "multiSimmap"</td></tr>
<tr><td style="width: 25%;"><a href="ctt.html">ctt</a></td>
<td>Generates (or simulates) a 'changes through time' plot from a set of stochastic map character histories</td></tr>
</table>

<h2><a name="D">-- D --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="anc.Bayes.html">density.anc.Bayes</a></td>
<td>Bayesian ancestral character estimation</td></tr>
<tr><td style="width: 25%;"><a href="fitMk.html">density.mcmcMk</a></td>
<td>Fits Mk model</td></tr>
<tr><td style="width: 25%;"><a href="density.multiSimmap.html">density.multiSimmap</a></td>
<td>Computes a posterior distribution for the number and types of changes on the tree</td></tr>
<tr><td style="width: 25%;"><a href="densityMap.html">densityMap</a></td>
<td>Plot posterior density of stochastic mapping on a tree</td></tr>
<tr><td style="width: 25%;"><a href="densityTree.html">densityTree</a></td>
<td>Plots a posterior sample of trees</td></tr>
<tr><td style="width: 25%;"><a href="describe.simmap.html">describe.simmap</a></td>
<td>Summarizes a stochastic mapped tree or set of trees</td></tr>
<tr><td style="width: 25%;"><a href="di2multi.simmap.html">di2multi.contMap</a></td>
<td>Collapse or resolve polytomies in a tree with a character painted on the edges</td></tr>
<tr><td style="width: 25%;"><a href="di2multi.simmap.html">di2multi.densityMap</a></td>
<td>Collapse or resolve polytomies in a tree with a character painted on the edges</td></tr>
<tr><td style="width: 25%;"><a href="di2multi.simmap.html">di2multi.multiSimmap</a></td>
<td>Collapse or resolve polytomies in a tree with a character painted on the edges</td></tr>
<tr><td style="width: 25%;"><a href="di2multi.simmap.html">di2multi.simmap</a></td>
<td>Collapse or resolve polytomies in a tree with a character painted on the edges</td></tr>
<tr><td style="width: 25%;"><a href="dotTree.html">dot.legend</a></td>
<td>Creates a phylogenetic dot plot</td></tr>
<tr><td style="width: 25%;"><a href="dotTree.html">dotTree</a></td>
<td>Creates a phylogenetic dot plot</td></tr>
<tr><td style="width: 25%;"><a href="drop.clade.html">drop.clade</a></td>
<td>Drop a clade from a tree</td></tr>
<tr><td style="width: 25%;"><a href="drop.leaves.html">drop.leaves</a></td>
<td>Drop all the leaves (tips) from a tree</td></tr>
<tr><td style="width: 25%;"><a href="drop.tip.contMap.html">drop.tip.contMap</a></td>
<td>Drop tip or tips from an object of class "contMap" or "densityMap"</td></tr>
<tr><td style="width: 25%;"><a href="drop.tip.contMap.html">drop.tip.densityMap</a></td>
<td>Drop tip or tips from an object of class "contMap" or "densityMap"</td></tr>
<tr><td style="width: 25%;"><a href="drop.tip.simmap.html">drop.tip.simmap</a></td>
<td>Drop tips or extract clade from tree with mapped discrete character</td></tr>
<tr><td style="width: 25%;"><a href="map.to.singleton.html">drop.tip.singleton</a></td>
<td>Converts a tree without singletons to a tree with singleton nodes</td></tr>
<tr><td style="width: 25%;"><a href="Dtest.html">Dtest</a></td>
<td>Conducts correlational D-test from stochastic mapping</td></tr>
</table>

<h2><a name="E">-- E --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="nodelabels.cophylo.html">edgelabels.cophylo</a></td>
<td>Add labels to a plotted "cophylo" object</td></tr>
<tr><td style="width: 25%;"><a href="edgeProbs.html">edgeProbs</a></td>
<td>Compute the relative frequencies of state changes along edges</td></tr>
<tr><td style="width: 25%;"><a href="contMap.html">errorbar.contMap</a></td>
<td>Map continuous trait evolution on the tree</td></tr>
<tr><td style="width: 25%;"><a href="estDiversity.html">estDiversity</a></td>
<td>Estimate diversity at each node of the tree</td></tr>
<tr><td style="width: 25%;"><a href="evol.rate.mcmc.html">evol.rate.mcmc</a></td>
<td>Bayesian MCMC method for identifying exceptional phenotypic diversification in a phylogeny</td></tr>
<tr><td style="width: 25%;"><a href="evol.vcv.html">evol.vcv</a></td>
<td>Likelihood test for variation in the evolutionary variance-covariance matrix</td></tr>
<tr><td style="width: 25%;"><a href="evolvcv.lite.html">evolvcv.lite</a></td>
<td>Likelihood test for a shift in the evolutionary correlation between traits</td></tr>
<tr><td style="width: 25%;"><a href="exhaustiveMP.html">exhaustiveMP</a></td>
<td>Exhaustive and branch &amp; bound MP optimization</td></tr>
<tr><td style="width: 25%;"><a href="expand.clade.html">expand.clade</a></td>
<td>Expands (or contracts) the tip-spacing of a given clade or clades</td></tr>
<tr><td style="width: 25%;"><a href="export.as.xml.html">export.as.xml</a></td>
<td>Export trees &amp; data in XML format</td></tr>
<tr><td style="width: 25%;"><a href="drop.tip.simmap.html">extract.clade.simmap</a></td>
<td>Drop tips or extract clade from tree with mapped discrete character</td></tr>
<tr><td style="width: 25%;"><a href="strahlerNumber.html">extract.strahlerNumber</a></td>
<td>Computes Strahler number for trees and nodes</td></tr>
</table>

<h2><a name="F">-- F --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="fancyTree.html">fancyTree</a></td>
<td>Plots special types of phylogenetic trees</td></tr>
<tr><td style="width: 25%;"><a href="fastAnc.html">fastAnc</a></td>
<td>(Reasonably) fast estimation of ML ancestral states</td></tr>
<tr><td style="width: 25%;"><a href="fastBM.html">fastBM</a></td>
<td>(Reasonably) fast quantitative trait simulation on phylogenies</td></tr>
<tr><td style="width: 25%;"><a href="fastMRCA.html">fastDist</a></td>
<td>Get the MRCA (or height above the root of the MRCA) of a pair of tip taxa</td></tr>
<tr><td style="width: 25%;"><a href="fastMRCA.html">fastHeight</a></td>
<td>Get the MRCA (or height above the root of the MRCA) of a pair of tip taxa</td></tr>
<tr><td style="width: 25%;"><a href="fastMRCA.html">fastMRCA</a></td>
<td>Get the MRCA (or height above the root of the MRCA) of a pair of tip taxa</td></tr>
<tr><td style="width: 25%;"><a href="findMRCA.html">findMRCA</a></td>
<td>Get the MRCA of a set of taxa</td></tr>
<tr><td style="width: 25%;"><a href="fit.bd.html">fit.bd</a></td>
<td>Fits birth-death (speciation/extinction) model to reconstructed phylogeny</td></tr>
<tr><td style="width: 25%;"><a href="fit.bd.html">fit.yule</a></td>
<td>Fits birth-death (speciation/extinction) model to reconstructed phylogeny</td></tr>
<tr><td style="width: 25%;"><a href="fitBayes.html">fitBayes</a></td>
<td>Evolutionary model fitting with intraspecific variability using Bayesian MCMC</td></tr>
<tr><td style="width: 25%;"><a href="fitDiversityModel.html">fitDiversityModel</a></td>
<td>Fit diversity-dependent phenotypic evolution model</td></tr>
<tr><td style="width: 25%;"><a href="fitMk.html">fitMk</a></td>
<td>Fits Mk model</td></tr>
<tr><td style="width: 25%;"><a href="fitMk.html">fitmultiMk</a></td>
<td>Fits Mk model</td></tr>
<tr><td style="width: 25%;"><a href="fitPagel.html">fitPagel</a></td>
<td>Function to test for correlated evolution of binary traits</td></tr>
<tr><td style="width: 25%;"><a href="fitMk.html">fitpolyMk</a></td>
<td>Fits Mk model</td></tr>
<tr><td style="width: 25%;"><a href="force.ultrametric.html">force.ultrametric</a></td>
<td>Forces a phylogenetic tree to be ultrametric</td></tr>
</table>

<h2><a name="G">-- G --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="gammatest.html">gammatest</a></td>
<td>Gamma test of Pybus &amp; Harvey (2000)</td></tr>
<tr><td style="width: 25%;"><a href="genSeq.html">genSeq</a></td>
<td>Simulate a DNA alignment on the tree under a model</td></tr>
<tr><td style="width: 25%;"><a href="add.species.to.genus.html">genus.to.species.tree</a></td>
<td>Add species to genus on a phylogeny or bind simulated species subtrees to a backbone genus tree</td></tr>
<tr><td style="width: 25%;"><a href="geo.legend.html">geo.legend</a></td>
<td>Adds a geological (or other temporal) legend to a plotted tree</td></tr>
<tr><td style="width: 25%;"><a href="geo.legend.html">geo.palette</a></td>
<td>Adds a geological (or other temporal) legend to a plotted tree</td></tr>
<tr><td style="width: 25%;"><a href="get.treepos.html">get.treepos</a></td>
<td>Get position or node of a plotted tree interactively</td></tr>
<tr><td style="width: 25%;"><a href="getCladesofSize.html">getCladesofSize</a></td>
<td>Get all subtrees larger than or equal to a specified size</td></tr>
<tr><td style="width: 25%;"><a href="getDescendants.html">getDescendants</a></td>
<td>Get descendant node numbers</td></tr>
<tr><td style="width: 25%;"><a href="getExtant.html">getExtant</a></td>
<td>Returns a list of the extant or extinct lineages in a tree containing non-contemporaneous tips</td></tr>
<tr><td style="width: 25%;"><a href="getExtant.html">getExtinct</a></td>
<td>Returns a list of the extant or extinct lineages in a tree containing non-contemporaneous tips</td></tr>
<tr><td style="width: 25%;"><a href="get.treepos.html">getnode</a></td>
<td>Get position or node of a plotted tree interactively</td></tr>
<tr><td style="width: 25%;"><a href="getDescendants.html">getParent</a></td>
<td>Get descendant node numbers</td></tr>
<tr><td style="width: 25%;"><a href="getSisters.html">getSisters</a></td>
<td>Get the sister node number, label, or set of nodes for a node or tip</td></tr>
<tr><td style="width: 25%;"><a href="getStates.html">getStates</a></td>
<td>Get the states at nodes or tips from a mapped tree</td></tr>
<tr><td style="width: 25%;"><a href="ltt.html">gtt</a></td>
<td>Creates lineage-through-time plot (including extinct lineages)</td></tr>
</table>

<h2><a name="L">-- L --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="labelnodes.html">labelnodes</a></td>
<td>Function to interactively label nodes of a plotted tree</td></tr>
<tr><td style="width: 25%;"><a href="ladderize.simmap.html">ladderize.simmap</a></td>
<td>Ladderize a tree with a mapped discrete character</td></tr>
<tr><td style="width: 25%;"><a href="lambda.transform.html">lambda.transform</a></td>
<td>lambda transformation of matrix</td></tr>
<tr><td style="width: 25%;"><a href="fit.bd.html">lik.bd</a></td>
<td>Fits birth-death (speciation/extinction) model to reconstructed phylogeny</td></tr>
<tr><td style="width: 25%;"><a href="likMlambda.html">likMlambda</a></td>
<td>Likelihood for joint lambda</td></tr>
<tr><td style="width: 25%;"><a href="rateshift.html">likSurface.rateshift</a></td>
<td>Find the temporal position of one or more rate shifts</td></tr>
<tr><td style="width: 25%;"><a href="linklabels.html">linklabels</a></td>
<td>Function to add tip labels to a plotted tree with linking lines</td></tr>
<tr><td style="width: 25%;"><a href="locate.fossil.html">locate.fossil</a></td>
<td>Locate a fossil lineage in a tree using continuous characters</td></tr>
<tr><td style="width: 25%;"><a href="locate.yeti.html">locate.yeti</a></td>
<td>Locate a cryptic, recently extinct, or missing taxon on a tree</td></tr>
<tr><td style="width: 25%;"><a href="fitDiversityModel.html">logLik.fitDiversityModel</a></td>
<td>Fit diversity-dependent phenotypic evolution model</td></tr>
<tr><td style="width: 25%;"><a href="averageTree.html">ls.consensus</a></td>
<td>Compute an average tree from a set of trees and related operations</td></tr>
<tr><td style="width: 25%;"><a href="ls.tree.html">ls.tree</a></td>
<td>Least squares branch lengths for a given tree</td></tr>
<tr><td style="width: 25%;"><a href="ltt.html">ltt</a></td>
<td>Creates lineage-through-time plot (including extinct lineages)</td></tr>
<tr><td style="width: 25%;"><a href="ltt95.html">ltt95</a></td>
<td>Creates a (1-alpha)-percent CI for a set of LTTs</td></tr>
</table>

<h2><a name="M">-- M --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="make.era.map.html">make.era.map</a></td>
<td>Create "era" map on a phylogenetic tree</td></tr>
<tr><td style="width: 25%;"><a href="make.simmap.html">make.simmap</a></td>
<td>Simulate stochastic character maps on a phylogenetic tree or trees</td></tr>
<tr><td style="width: 25%;"><a href="densityTree.html">make.transparent</a></td>
<td>Plots a posterior sample of trees</td></tr>
<tr><td style="width: 25%;"><a href="anoletree.html">mammal.data</a></td>
<td>Phylogenetic datasets</td></tr>
<tr><td style="width: 25%;"><a href="anoletree.html">mammal.tree</a></td>
<td>Phylogenetic datasets</td></tr>
<tr><td style="width: 25%;"><a href="map.overlap.html">Map.Overlap</a></td>
<td>Proportional overlap between two mapped character histories on a tree</td></tr>
<tr><td style="width: 25%;"><a href="map.overlap.html">map.overlap</a></td>
<td>Proportional overlap between two mapped character histories on a tree</td></tr>
<tr><td style="width: 25%;"><a href="map.to.singleton.html">map.to.singleton</a></td>
<td>Converts a tree without singletons to a tree with singleton nodes</td></tr>
<tr><td style="width: 25%;"><a href="mapped.states.html">mapped.states</a></td>
<td>Returns a vector, matrix, or list of the mapped states on a tree or set of trees</td></tr>
<tr><td style="width: 25%;"><a href="markChanges.html">markChanges</a></td>
<td>Add marked changes to a plotted tree with mapped discrete character</td></tr>
<tr><td style="width: 25%;"><a href="matchNodes.html">matchLabels</a></td>
<td>Matches nodes between two trees</td></tr>
<tr><td style="width: 25%;"><a href="matchNodes.html">matchNodes</a></td>
<td>Matches nodes between two trees</td></tr>
<tr><td style="width: 25%;"><a href="ltt.html">mccr</a></td>
<td>Creates lineage-through-time plot (including extinct lineages)</td></tr>
<tr><td style="width: 25%;"><a href="fitMk.html">mcmcMk</a></td>
<td>Fits Mk model</td></tr>
<tr><td style="width: 25%;"><a href="mergeMappedStates.html">mergeMappedStates</a></td>
<td>Merge two or more mapped states into one state</td></tr>
<tr><td style="width: 25%;"><a href="midpoint.root.html">midpoint.root</a></td>
<td>Midpoint root a phylogeny</td></tr>
<tr><td style="width: 25%;"><a href="minRotate.html">minRotate</a></td>
<td>Rotates all nodes of the tree to minimize the difference in order with a vector</td></tr>
<tr><td style="width: 25%;"><a href="minSplit.html">minSplit</a></td>
<td>Finding the minimum (median) split in the posterior sample</td></tr>
<tr><td style="width: 25%;"><a href="averageTree.html">minTreeDist</a></td>
<td>Compute an average tree from a set of trees and related operations</td></tr>
<tr><td style="width: 25%;"><a href="modified.Grafen.html">modified.Grafen</a></td>
<td>Computes modified Grafen edge lengths</td></tr>
<tr><td style="width: 25%;"><a href="mrp.supertree.html">mrp.supertree</a></td>
<td>Matrix representation parsimony supertree estimation</td></tr>
<tr><td style="width: 25%;"><a href="multi.mantel.html">multi.mantel</a></td>
<td>Multiple matrix regression (partial Mantel test)</td></tr>
<tr><td style="width: 25%;"><a href="di2multi.simmap.html">multi2di.contMap</a></td>
<td>Collapse or resolve polytomies in a tree with a character painted on the edges</td></tr>
<tr><td style="width: 25%;"><a href="di2multi.simmap.html">multi2di.densityMap</a></td>
<td>Collapse or resolve polytomies in a tree with a character painted on the edges</td></tr>
<tr><td style="width: 25%;"><a href="di2multi.simmap.html">multi2di.multiSimmap</a></td>
<td>Collapse or resolve polytomies in a tree with a character painted on the edges</td></tr>
<tr><td style="width: 25%;"><a href="di2multi.simmap.html">multi2di.simmap</a></td>
<td>Collapse or resolve polytomies in a tree with a character painted on the edges</td></tr>
<tr><td style="width: 25%;"><a href="multiC.html">multiC</a></td>
<td>Returns a list with phylogenetic VCV matrix for each mapped state</td></tr>
<tr><td style="width: 25%;"><a href="sim.rates.html">multiOU</a></td>
<td>Brownian or OU simulation with multiple evolutionary regimes</td></tr>
<tr><td style="width: 25%;"><a href="multiRF.html">multiRF</a></td>
<td>Computes Robinson-Foulds distance between a set of trees</td></tr>
</table>

<h2><a name="N">-- N --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="modified.Grafen.html">node.paths</a></td>
<td>Computes modified Grafen edge lengths</td></tr>
<tr><td style="width: 25%;"><a href="nodeHeights.html">nodeheight</a></td>
<td>Compute the heights above the root of each node</td></tr>
<tr><td style="width: 25%;"><a href="nodeHeights.html">nodeHeights</a></td>
<td>Compute the heights above the root of each node</td></tr>
<tr><td style="width: 25%;"><a href="nodelabels.cophylo.html">nodelabels.cophylo</a></td>
<td>Add labels to a plotted "cophylo" object</td></tr>
</table>

<h2><a name="O">-- O --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="optim.phylo.ls.html">optim.phylo.ls</a></td>
<td>Phylogeny inference using the least squares method</td></tr>
<tr><td style="width: 25%;"><a href="orderMappedEdge.html">orderMappedEdge</a></td>
<td>Order the columns of mapped.edge to match across trees</td></tr>
</table>

<h2><a name="P">-- P --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="paintSubTree.html">paintBranches</a></td>
<td>Paint sub-trees with a discrete character</td></tr>
<tr><td style="width: 25%;"><a href="paintSubTree.html">paintSubTree</a></td>
<td>Paint sub-trees with a discrete character</td></tr>
<tr><td style="width: 25%;"><a href="paste.tree.html">paste.tree</a></td>
<td>Paste two trees together</td></tr>
<tr><td style="width: 25%;"><a href="pbtree.html">pbtree</a></td>
<td>Simulate pure-birth or birth-death stochastic tree or trees</td></tr>
<tr><td style="width: 25%;"><a href="pgls.Ives.html">pgls.Ives</a></td>
<td>Phylogenetic regression with intraspecific sampling error</td></tr>
<tr><td style="width: 25%;"><a href="pgls.Ives.html">pgls.SEy</a></td>
<td>Phylogenetic regression with intraspecific sampling error</td></tr>
<tr><td style="width: 25%;"><a href="phenogram.html">phenogram</a></td>
<td>Plot phenogram (traitgram)</td></tr>
<tr><td style="width: 25%;"><a href="fancyTree.html">phenogram95</a></td>
<td>Plots special types of phylogenetic trees</td></tr>
<tr><td style="width: 25%;"><a href="phyl.cca.html">phyl.cca</a></td>
<td>Phylogenetic canonical correlation analysis</td></tr>
<tr><td style="width: 25%;"><a href="phyl.pairedttest.html">phyl.pairedttest</a></td>
<td>Phylogenetic paired t-test</td></tr>
<tr><td style="width: 25%;"><a href="phyl.pca.html">phyl.pca</a></td>
<td>Phylogenetic principal components analysis</td></tr>
<tr><td style="width: 25%;"><a href="phyl.resid.html">phyl.resid</a></td>
<td>Phylogenetic size-correction via GLS regression</td></tr>
<tr><td style="width: 25%;"><a href="phyl.RMA.html">phyl.RMA</a></td>
<td>Phylogenetic reduced major axis (RMA) regression</td></tr>
<tr><td style="width: 25%;"><a href="phyl.vcv.html">phyl.vcv</a></td>
<td>Compute evolutionary VCV matrix for a tree &amp; dataset</td></tr>
<tr><td style="width: 25%;"><a href="phylANOVA.html">phylANOVA</a></td>
<td>Phylogenetic ANOVA and post-hoc tests</td></tr>
<tr><td style="width: 25%;"><a href="phylo.heatmap.html">phylo.heatmap</a></td>
<td>Creates a phylogenetic heat map</td></tr>
<tr><td style="width: 25%;"><a href="phylo.impute.html">phylo.impute</a></td>
<td>Phylogenetic imputation for multivariate continuous character data</td></tr>
<tr><td style="width: 25%;"><a href="phylo.to.map.html">phylo.to.map</a></td>
<td>Plot tree with tips linked to geographic coordinates</td></tr>
<tr><td style="width: 25%;"><a href="phylo.toBackbone.html">phylo.toBackbone</a></td>
<td>Converts tree to backbone or vice versa</td></tr>
<tr><td style="width: 25%;"><a href="phyloDesign.html">phyloDesign</a></td>
<td>Compute design matrix for least squares analyses</td></tr>
<tr><td style="width: 25%;"><a href="phylomorphospace.html">phylomorphospace</a></td>
<td>Creates phylomorphospace plot</td></tr>
<tr><td style="width: 25%;"><a href="phylomorphospace3d.html">phylomorphospace3d</a></td>
<td>Creates tree-dimensional phylomorphospace plot</td></tr>
<tr><td style="width: 25%;"><a href="fancyTree.html">phyloScattergram</a></td>
<td>Plots special types of phylogenetic trees</td></tr>
<tr><td style="width: 25%;"><a href="phylosig.html">phylosig</a></td>
<td>Compute phylogenetic signal with two methods</td></tr>
<tr><td style="width: 25%;"><a href="phytools-package.html">phytools</a></td>
<td>phytools: Phylogenetic Tools for comparative biology (and other things)</td></tr>
<tr><td style="width: 25%;"><a href="anc.Bayes.html">plot.anc.Bayes</a></td>
<td>Bayesian ancestral character estimation</td></tr>
<tr><td style="width: 25%;"><a href="plot.backbonePhylo.html">plot.backbonePhylo</a></td>
<td>Plots backbone tree with triangles as clades</td></tr>
<tr><td style="width: 25%;"><a href="density.multiSimmap.html">plot.changesMap</a></td>
<td>Computes a posterior distribution for the number and types of changes on the tree</td></tr>
<tr><td style="width: 25%;"><a href="contMap.html">plot.contMap</a></td>
<td>Map continuous trait evolution on the tree</td></tr>
<tr><td style="width: 25%;"><a href="cophylo.html">plot.cophylo</a></td>
<td>Creates a co-phylogenetic plot</td></tr>
<tr><td style="width: 25%;"><a href="cospeciation.html">plot.cospeciation</a></td>
<td>Conducts a statistical test of cospeciation between two trees</td></tr>
<tr><td style="width: 25%;"><a href="fitMk.html">plot.density.mcmcMk</a></td>
<td>Fits Mk model</td></tr>
<tr><td style="width: 25%;"><a href="densityMap.html">plot.densityMap</a></td>
<td>Plot posterior density of stochastic mapping on a tree</td></tr>
<tr><td style="width: 25%;"><a href="describe.simmap.html">plot.describe.simmap</a></td>
<td>Summarizes a stochastic mapped tree or set of trees</td></tr>
<tr><td style="width: 25%;"><a href="expand.clade.html">plot.expand.clade</a></td>
<td>Expands (or contracts) the tip-spacing of a given clade or clades</td></tr>
<tr><td style="width: 25%;"><a href="fitMk.html">plot.fitMk</a></td>
<td>Fits Mk model</td></tr>
<tr><td style="width: 25%;"><a href="fitPagel.html">plot.fitPagel</a></td>
<td>Function to test for correlated evolution of binary traits</td></tr>
<tr><td style="width: 25%;"><a href="fitMk.html">plot.fitpolyMk</a></td>
<td>Fits Mk model</td></tr>
<tr><td style="width: 25%;"><a href="fitMk.html">plot.gfit</a></td>
<td>Fits Mk model</td></tr>
<tr><td style="width: 25%;"><a href="ltt95.html">plot.ltt95</a></td>
<td>Creates a (1-alpha)-percent CI for a set of LTTs</td></tr>
<tr><td style="width: 25%;"><a href="fitMk.html">plot.mcmcMk</a></td>
<td>Fits Mk model</td></tr>
<tr><td style="width: 25%;"><a href="plotSimmap.html">plot.multiSimmap</a></td>
<td>Plot stochastic character mapped tree</td></tr>
<tr><td style="width: 25%;"><a href="phyl.RMA.html">plot.phyl.RMA</a></td>
<td>Phylogenetic reduced major axis (RMA) regression</td></tr>
<tr><td style="width: 25%;"><a href="phylo.to.map.html">plot.phylo.to.map</a></td>
<td>Plot tree with tips linked to geographic coordinates</td></tr>
<tr><td style="width: 25%;"><a href="phylosig.html">plot.phylosig</a></td>
<td>Compute phylogenetic signal with two methods</td></tr>
<tr><td style="width: 25%;"><a href="as.Qmatrix.html">plot.Qmatrix</a></td>
<td>Convert a fitted M_k_ model to a Q-matrix</td></tr>
<tr><td style="width: 25%;"><a href="rateshift.html">plot.rateshift</a></td>
<td>Find the temporal position of one or more rate shifts</td></tr>
<tr><td style="width: 25%;"><a href="simBMphylo.html">plot.simBMphylo</a></td>
<td>Creates a graphical illustration of Brownian motion evolution on a phylogeny</td></tr>
<tr><td style="width: 25%;"><a href="plotSimmap.html">plot.simmap</a></td>
<td>Plot stochastic character mapped tree</td></tr>
<tr><td style="width: 25%;"><a href="evol.rate.mcmc.html">plot.summary.evol.rate.mcmc</a></td>
<td>Bayesian MCMC method for identifying exceptional phenotypic diversification in a phylogeny</td></tr>
<tr><td style="width: 25%;"><a href="plotBranchbyTrait.html">plotBranchbyTrait</a></td>
<td>Plot branch colors by a quantitative trait or value</td></tr>
<tr><td style="width: 25%;"><a href="plotSimmap.html">plotSimmap</a></td>
<td>Plot stochastic character mapped tree</td></tr>
<tr><td style="width: 25%;"><a href="plotThresh.html">plotThresh</a></td>
<td>Tree plotting with posterior probabilities of ancestral states from the threshold model</td></tr>
<tr><td style="width: 25%;"><a href="plotTree.html">plotTree</a></td>
<td>Plots rooted phylogenetic tree</td></tr>
<tr><td style="width: 25%;"><a href="plotTree.wBars.html">plotTree.barplot</a></td>
<td>Plot a tree with bars at the tips</td></tr>
<tr><td style="width: 25%;"><a href="plotTree.wBars.html">plotTree.boxplot</a></td>
<td>Plot a tree with bars at the tips</td></tr>
<tr><td style="width: 25%;"><a href="plotTree.datamatrix.html">plotTree.datamatrix</a></td>
<td>Plot a tree with a discrete character data matrix at the tips</td></tr>
<tr><td style="width: 25%;"><a href="plotTree.errorbars.html">plotTree.errorbars</a></td>
<td>Plot a tree with error bars around divergence dates</td></tr>
<tr><td style="width: 25%;"><a href="map.to.singleton.html">plotTree.singletons</a></td>
<td>Converts a tree without singletons to a tree with singleton nodes</td></tr>
<tr><td style="width: 25%;"><a href="splitplotTree.html">plotTree.splits</a></td>
<td>Plots a phylogeny in two columns</td></tr>
<tr><td style="width: 25%;"><a href="plotTree.wBars.html">plotTree.wBars</a></td>
<td>Plot a tree with bars at the tips</td></tr>
<tr><td style="width: 25%;"><a href="posterior.evolrate.html">posterior.evolrate</a></td>
<td>Analysis of the posterior sample from evol.rate.mcmc</td></tr>
<tr><td style="width: 25%;"><a href="posthoc.html">posthoc</a></td>
<td>Generic post-hoc test</td></tr>
<tr><td style="width: 25%;"><a href="ratebytree.html">posthoc.ratebytree</a></td>
<td>Likelihood test for rate variation among trees, clades, or traits</td></tr>
<tr><td style="width: 25%;"><a href="print.backbonePhylo.html">print.backbonePhylo</a></td>
<td>Print method for backbone phylogeny</td></tr>
<tr><td style="width: 25%;"><a href="cospeciation.html">print.cospeciation</a></td>
<td>Conducts a statistical test of cospeciation between two trees</td></tr>
<tr><td style="width: 25%;"><a href="evol.rate.mcmc.html">print.evol.rate.mcmc</a></td>
<td>Bayesian MCMC method for identifying exceptional phenotypic diversification in a phylogeny</td></tr>
<tr><td style="width: 25%;"><a href="fit.bd.html">print.fit.bd</a></td>
<td>Fits birth-death (speciation/extinction) model to reconstructed phylogeny</td></tr>
<tr><td style="width: 25%;"><a href="fitDiversityModel.html">print.fitDiversityModel</a></td>
<td>Fit diversity-dependent phenotypic evolution model</td></tr>
<tr><td style="width: 25%;"><a href="as.Qmatrix.html">print.Qmatrix</a></td>
<td>Convert a fitted M_k_ model to a Q-matrix</td></tr>
<tr><td style="width: 25%;"><a href="evol.rate.mcmc.html">print.summary.evol.rate.mcmc</a></td>
<td>Bayesian MCMC method for identifying exceptional phenotypic diversification in a phylogeny</td></tr>
<tr><td style="width: 25%;"><a href="phylomorphospace.html">project.phylomorphospace</a></td>
<td>Creates phylomorphospace plot</td></tr>
</table>

<h2><a name="R">-- R --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="ratebystate.html">ratebystate</a></td>
<td>Method for investigating the rate of one trait as a function of the state of another</td></tr>
<tr><td style="width: 25%;"><a href="ratebytree.html">ratebytree</a></td>
<td>Likelihood test for rate variation among trees, clades, or traits</td></tr>
<tr><td style="width: 25%;"><a href="rateshift.html">rateshift</a></td>
<td>Find the temporal position of one or more rate shifts</td></tr>
<tr><td style="width: 25%;"><a href="read.newick.html">read.newick</a></td>
<td>Robust Newick style tree reader</td></tr>
<tr><td style="width: 25%;"><a href="read.simmap.html">read.simmap</a></td>
<td>Read SIMMAP style trees from file</td></tr>
<tr><td style="width: 25%;"><a href="reorder.backbonePhylo.html">reorder.backbonePhylo</a></td>
<td>Reorders a backbone phylogeny</td></tr>
<tr><td style="width: 25%;"><a href="reorderSimmap.html">reorderSimmap</a></td>
<td>Reorder edges of a simmap tree</td></tr>
<tr><td style="width: 25%;"><a href="rep.phylo.html">rep.multiPhylo</a></td>
<td>Replicate a tree or set of trees</td></tr>
<tr><td style="width: 25%;"><a href="rep.phylo.html">rep.phylo</a></td>
<td>Replicate a tree or set of trees</td></tr>
<tr><td style="width: 25%;"><a href="rep.phylo.html">repPhylo</a></td>
<td>Replicate a tree or set of trees</td></tr>
<tr><td style="width: 25%;"><a href="reroot.html">reroot</a></td>
<td>Re-root a tree along an edge</td></tr>
<tr><td style="width: 25%;"><a href="rerootingMethod.html">rerootingMethod</a></td>
<td>Get marginal ancestral state reconstructions by re-rooting</td></tr>
<tr><td style="width: 25%;"><a href="rescaleSimmap.html">rescaleSimmap</a></td>
<td>Rescale SIMMAP style tree</td></tr>
<tr><td style="width: 25%;"><a href="resolveNode.html">resolveAllNodes</a></td>
<td>Compute all possible resolutions of a node or all nodes in a multifurcating tree</td></tr>
<tr><td style="width: 25%;"><a href="resolveNode.html">resolveNode</a></td>
<td>Compute all possible resolutions of a node or all nodes in a multifurcating tree</td></tr>
<tr><td style="width: 25%;"><a href="map.to.singleton.html">rootedge.to.singleton</a></td>
<td>Converts a tree without singletons to a tree with singleton nodes</td></tr>
<tr><td style="width: 25%;"><a href="rotateNodes.html">rotate.multi</a></td>
<td>Rotates a node or set of nodes in a phylogenetic tree</td></tr>
<tr><td style="width: 25%;"><a href="rotateNodes.html">rotateNodes</a></td>
<td>Rotates a node or set of nodes in a phylogenetic tree</td></tr>
<tr><td style="width: 25%;"><a href="roundBranches.html">roundBranches</a></td>
<td>Rounds the branch lengths of a tree</td></tr>
<tr><td style="width: 25%;"><a href="roundPhylogram.html">roundPhylogram</a></td>
<td>Plot a round phylogram</td></tr>
<tr><td style="width: 25%;"><a href="rstate.html">rstate</a></td>
<td>Pick a random state according to a vector of probabilities</td></tr>
</table>

<h2><a name="S">-- S --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="anoletree.html">salamanders</a></td>
<td>Phylogenetic datasets</td></tr>
<tr><td style="width: 25%;"><a href="sampleFrom.html">sampleFrom</a></td>
<td>Sample from a set of distributions</td></tr>
<tr><td style="width: 25%;"><a href="phyl.pca.html">scores</a></td>
<td>Phylogenetic principal components analysis</td></tr>
<tr><td style="width: 25%;"><a href="phyl.pca.html">scores.phyl.pca</a></td>
<td>Phylogenetic principal components analysis</td></tr>
<tr><td style="width: 25%;"><a href="setMap.html">setMap</a></td>
<td>Set color map for various phylogenetic objects of classes</td></tr>
<tr><td style="width: 25%;"><a href="setMap.html">setMap.contMap</a></td>
<td>Set color map for various phylogenetic objects of classes</td></tr>
<tr><td style="width: 25%;"><a href="setMap.html">setMap.densityMap</a></td>
<td>Set color map for various phylogenetic objects of classes</td></tr>
<tr><td style="width: 25%;"><a href="setMap.html">setMap.phyloScattergram</a></td>
<td>Set color map for various phylogenetic objects of classes</td></tr>
<tr><td style="width: 25%;"><a href="sim.corrs.html">sim.corrs</a></td>
<td>Multivariate Brownian simulation with multiple correlations and rates</td></tr>
<tr><td style="width: 25%;"><a href="ctt.html">sim.ctt</a></td>
<td>Generates (or simulates) a 'changes through time' plot from a set of stochastic map character histories</td></tr>
<tr><td style="width: 25%;"><a href="sim.history.html">sim.history</a></td>
<td>Simulate character history or a discrete character at the tips of the tree under some model</td></tr>
<tr><td style="width: 25%;"><a href="sim.history.html">sim.Mk</a></td>
<td>Simulate character history or a discrete character at the tips of the tree under some model</td></tr>
<tr><td style="width: 25%;"><a href="ctt.html">sim.multiCtt</a></td>
<td>Generates (or simulates) a 'changes through time' plot from a set of stochastic map character histories</td></tr>
<tr><td style="width: 25%;"><a href="sim.history.html">sim.multiMk</a></td>
<td>Simulate character history or a discrete character at the tips of the tree under some model</td></tr>
<tr><td style="width: 25%;"><a href="sim.ratebystate.html">sim.ratebystate</a></td>
<td>Conduct simulation of state dependent rate variation</td></tr>
<tr><td style="width: 25%;"><a href="sim.rates.html">sim.rates</a></td>
<td>Brownian or OU simulation with multiple evolutionary regimes</td></tr>
<tr><td style="width: 25%;"><a href="simBMphylo.html">simBMphylo</a></td>
<td>Creates a graphical illustration of Brownian motion evolution on a phylogeny</td></tr>
<tr><td style="width: 25%;"><a href="skewers.html">skewers</a></td>
<td>Matrix comparison using the method of random skewers</td></tr>
<tr><td style="width: 25%;"><a href="splitEdgeColor.html">splitEdgeColor</a></td>
<td>Split edge colors when descendant edges have different mapped states</td></tr>
<tr><td style="width: 25%;"><a href="splitplotTree.html">splitplotTree</a></td>
<td>Plots a phylogeny in two columns</td></tr>
<tr><td style="width: 25%;"><a href="splitTree.html">splitTree</a></td>
<td>Split tree at a point</td></tr>
<tr><td style="width: 25%;"><a href="starTree.html">starTree</a></td>
<td>Create star phylogeny</td></tr>
<tr><td style="width: 25%;"><a href="strahlerNumber.html">strahlerNumber</a></td>
<td>Computes Strahler number for trees and nodes</td></tr>
<tr><td style="width: 25%;"><a href="evol.rate.mcmc.html">summary.evol.rate.mcmc</a></td>
<td>Bayesian MCMC method for identifying exceptional phenotypic diversification in a phylogeny</td></tr>
<tr><td style="width: 25%;"><a href="describe.simmap.html">summary.multiSimmap</a></td>
<td>Summarizes a stochastic mapped tree or set of trees</td></tr>
<tr><td style="width: 25%;"><a href="describe.simmap.html">summary.simmap</a></td>
<td>Summarizes a stochastic mapped tree or set of trees</td></tr>
<tr><td style="width: 25%;"><a href="anoletree.html">sunfish.data</a></td>
<td>Phylogenetic datasets</td></tr>
<tr><td style="width: 25%;"><a href="anoletree.html">sunfish.tree</a></td>
<td>Phylogenetic datasets</td></tr>
</table>

<h2><a name="T">-- T --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="threshBayes.html">threshBayes</a></td>
<td>Threshold model using Bayesian MCMC</td></tr>
<tr><td style="width: 25%;"><a href="threshDIC.html">threshDIC</a></td>
<td>Deviance Information Criterion from the threshold model</td></tr>
<tr><td style="width: 25%;"><a href="threshState.html">threshState</a></td>
<td>Computes value for a threshold character from a liability and thresholds</td></tr>
<tr><td style="width: 25%;"><a href="nodelabels.cophylo.html">tiplabels.cophylo</a></td>
<td>Add labels to a plotted "cophylo" object</td></tr>
<tr><td style="width: 25%;"><a href="minRotate.html">tipRotate</a></td>
<td>Rotates all nodes of the tree to minimize the difference in order with a vector</td></tr>
<tr><td style="width: 25%;"><a href="to.matrix.html">to.matrix</a></td>
<td>Convert a character vector to a binary matrix</td></tr>
<tr><td style="width: 25%;"><a href="tree.grow.html">tree.grow</a></td>
<td>Creates an animation of a tree growing from left-to-right or upwards</td></tr>
<tr><td style="width: 25%;"><a href="treeSlice.html">treeSlice</a></td>
<td>Slices the tree at a particular point and returns all subtrees, or the tree rootward of the point</td></tr>
</table>

<h2><a name="U">-- U --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="untangle.html">untangle</a></td>
<td>Attempts to untangle crossing branches for plotting</td></tr>
</table>

<h2><a name="V">-- V --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="vcvPhylo.html">vcvPhylo</a></td>
<td>Calculates cophenetic (i.e., phylogenetic VCV) matrix</td></tr>
</table>

<h2><a name="W">-- W --</a></h2>

<table width="100%">
<tr><td style="width: 25%;"><a href="anoletree.html">wasp.data</a></td>
<td>Phylogenetic datasets</td></tr>
<tr><td style="width: 25%;"><a href="anoletree.html">wasp.trees</a></td>
<td>Phylogenetic datasets</td></tr>
<tr><td style="width: 25%;"><a href="write.simmap.html">write.simmap</a></td>
<td>Write a stochastic character mapped tree to file</td></tr>
<tr><td style="width: 25%;"><a href="writeAncestors.html">writeAncestors</a></td>
<td>Write a tree to file with ancestral states and (optionally) CIs at nodes</td></tr>
<tr><td style="width: 25%;"><a href="writeNexus.html">writeNexus</a></td>
<td>Write a tree to file in Nexus format</td></tr>
</table>
</body></html>
